File: src/main/java/frc/robot/subsystems/superstructure/arm/Arm.java
package frc.robot.subsystems.superstructure.arm;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.units.Units;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine;
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine.Direction;
import frc.robot.Constants.ArmConstants;
import frc.robot.RobotType;
import frc.robot.subsystems.superstructure.arm.ArmIOInputsAutoLogged;
import java.util.function.Supplier;
import org.littletonrobotics.junction.Logger;

/** Arm subsystem wrapper around IO implementations. */
public final class Arm extends SubsystemBase {
  private final ArmIO io;
  private final ArmIOInputsAutoLogged inputs = new ArmIOInputsAutoLogged();
  private final SysIdRoutine sysIdRoutine;

  public Arm() {
    this.io =
        switch (RobotType.MODE) {
          case REAL -> new ArmIOKraken();
          case SIMULATION -> new ArmIOSim();
          case REPLAY -> new ArmIO() {};
        };
    
    this.sysIdRoutine = new SysIdRoutine(
        new SysIdRoutine.Config(
            null,
            Units.Volts.of(1.0),
            null,
            state -> Logger.recordOutput("arm/sysid", state.toString())),
        new SysIdRoutine.Mechanism(voltage -> io.setVoltage(voltage.in(Units.Volts)), null, this));
  }

  @Override
  public void periodic() {
    io.updateInputs(inputs);
    Logger.processInputs("Arm", inputs);
    Logger.recordOutput("Arm/AngleRad", inputs.angle.getRadians());
    Logger.recordOutput("Arm/VelocityRadPerSec", inputs.angularVelocityRadPerSec);
    Logger.recordOutput("Arm/AppliedVolts", inputs.appliedVolts);
  }

  private Rotation2d clampAngle(Rotation2d angle) {
    double min = ArmConstants.MIN_ANGLE.getRadians();
    double max = ArmConstants.MAX_ANGLE.getRadians();
    double clamped = Math.max(min, Math.min(max, angle.getRadians()));
    return Rotation2d.fromRadians(clamped);
  }

  public void setBrakeMode(boolean brake) {
    io.setBrakeMode(brake);
  }

  public void setVoltage(double volts) {
    io.setVoltage(volts);
  }

  public void stop() {
    io.stop();
  }

  public void setTargetAngle(Rotation2d angle) {
    io.setTargetAngle(clampAngle(angle));
  }

  public Command getMoveToAngleCommand(Supplier<Rotation2d> angleSupplier) {
    return Commands.runOnce(() -> setTargetAngle(angleSupplier.get()));
  }

  public Command getDynamicSysID(Direction direction) {
    return sysIdRoutine.dynamic(direction);
  }

  public Command getQuasistaticSysID(Direction direction) {
    return sysIdRoutine.quasistatic(direction);
  }

  public ArmIO.ArmIOInputs getInputs() {
    return inputs;
  }

  public Rotation2d getAngle() {
    return inputs.angle;
  }
}

File: src/main/java/frc/robot/subsystems/superstructure/arm/Arm.kt
package frc.robot.subsystems.superstructure.arm

import edu.wpi.first.units.Units.*
import edu.wpi.first.units.measure.*
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.Commands.runOnce
import edu.wpi.first.wpilibj2.command.SubsystemBase
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine.Direction
import frc.robot.Constants.ArmConstants
import frc.robot.RobotType
import org.littletonrobotics.junction.Logger
import java.util.function.Supplier

class Arm : SubsystemBase() {
    val io: ArmIO = when (RobotType.mode) {
        RobotType.Mode.REAL -> ArmIOReal()
        RobotType.Mode.SIMULATION -> ArmIOSim()
        RobotType.Mode.REPLAY -> object : ArmIO {}
    }

    val inputs = ArmIO.ArmInputs()

    private val sysIdRoutine: SysIdRoutine = SysIdRoutine(
        SysIdRoutine.Config(
            null,
            Volts.of(4.0),
            null,
            { state -> Logger.recordOutput("arm/sysid", state.toString()) }
        ),
        SysIdRoutine.Mechanism(
            { voltage: Voltage -> io.setVoltage(voltage) },
            null,
            this
        )
    )

    override fun periodic() {
        io.updateInputs(inputs)
        Logger.processInputs("Arm", inputs)

        val angleRad = inputs.angle.baseUnitMagnitude()
        val velocityRadPerSec = inputs.angularVelocity.baseUnitMagnitude()
        val applied = inputs.appliedVoltage.baseUnitMagnitude()

        Logger.recordOutput("Arm/AngleRad", angleRad)
        Logger.recordOutput("Arm/VelocityRadPerSec", velocityRadPerSec)
        Logger.recordOutput("Arm/AppliedVolts", applied)

    }

    private fun clampAngle(angle: Angle): Angle {
        val min = ArmConstants.MIN_ANGLE.baseUnitMagnitude()
        val max = ArmConstants.MAX_ANGLE.baseUnitMagnitude()
        val rad = angle.baseUnitMagnitude()
        val clamped = rad.coerceIn(min, max)
        return Radians.of(clamped)
    }

    fun setBrakeMode(brake: Boolean) { io.setBrakeMode(brake) }

    fun setVoltage(voltage: Voltage) { io.setVoltage(voltage) }

    fun stop() { io.stop() }

    fun setTargetAngle(angle: Angle) { io.setTargetAngle(clampAngle(angle)) }

    fun getMoveToAngleCommand(angleSupplier: Supplier<Angle>): Command {
        return runOnce { setTargetAngle(angleSupplier.get()) }
    }

    fun getDynamicSysID(direction: Direction): Command = sysIdRoutine.dynamic(direction)
    fun getQuasistaticSysID(direction: Direction): Command = sysIdRoutine.quasistatic(direction)
}

